pkg/controllers/jobset_controller.go

```diff
@@ -951,9 +881,159 @@ func findJobFailureTime(job *batchv1.Job) *metav1.Time {
+// the JobSet, if one exists. Otherwise, it returns nil.
+func managedByExternalController(js *jobset.JobSet) *string {
 	if controllerName := js.Spec.ManagedBy; controllerName != nil && *controllerName != jobset.JobSetControllerName {
 		return controllerName
 	}
 	return nil
 }
+
+// enqueueEvent appends a new k8s event to be emitted if and only after running the status
+// update functions in the updateStatusOpts, the status update API call suceeds.
+func enqueueEvent(updateStatusOpts *statusUpdateOpts, event *eventParams) {
+	updateStatusOpts.events = append(updateStatusOpts.events, event)
+}
+
+// function parameters for setCondition
+type conditionOpts struct {
+	eventType string
+	condition *metav1.Condition
+}
+
+// setCondition will add a new condition to the JobSet status (or update an existing one),
+// and enqueue an event for emission if the status update succeeds at the end of the reconcile.
+func setCondition(js *jobset.JobSet, condOpts *conditionOpts, updateStatusOpts *statusUpdateOpts) {
+	// Return early if this condition is already set.
+	if !updateCondition(js, condOpts) {
+		return
+	}
+
+	// If we made some changes to the status conditions, configure updateStatusOpts
+	// to persist the status update at the end of the reconciliation attempt.
+	updateStatusOpts.shouldUpdate = true
+
+	// Conditionally emit an event for each JobSet condition update if and only if
+	// the status update call is successful.
+	enqueueEvent(updateStatusOpts, &eventParams{
+		object:       js,
+		eventType:    condOpts.eventType,
+		eventReason:  condOpts.condition.Reason,
+		eventMessage: condOpts.condition.Message,
+	})
+}
+
+// updateCondition accepts a given condition and does one of the following:
+//  1. If an identical condition already exists, do nothing and return false (indicating
+//     no change was made).
+//  2. If a condition of the same type exists but with a different status, update
+//     the condition in place and return true (indicating a condition change was made).
+func updateCondition(js *jobset.JobSet, opts *conditionOpts) bool {
+	if opts == nil || opts.condition == nil {
+		return false
+	}
+
+	condition := *opts.condition
+	condition.LastTransitionTime = metav1.Now()
+	for i, val := range js.Status.Conditions {
+		if condition.Type == val.Type && condition.Status != val.Status {
+			js.Status.Conditions[i] = condition
+			// Condition found but different status so we should update.
+			return true
+		} else if condition.Type == val.Type && condition.Status == val.Status && condition.Reason == val.Reason && condition.Message == val.Message {
+			// Duplicate condition so no update.
+			return false
+		}
+	}
+
+	// Condition doesn't exist, add it.
+	js.Status.Conditions = append(js.Status.Conditions, condition)
+	return true
+}
+
+// setJobSetCompleted sets a condition on the JobSet status indicating it has completed.
+func setJobSetCompleted(js *jobset.JobSet, updateStatusOpts *statusUpdateOpts) {
```

Comment:
I would suffix this and the functions below with `Condition`

Reply:
Done